### PR TITLE
Added ToHashSetAsync<T> Extension methods on IQueryable<T>

### LINKED ISSUE #30033: Add ToHashSetAsync() to IQueryable<T>
In current EF Core code, most queries are written like this
```
List<Blog> blogs = await context.Blogs.Where(b => something).ToListAsync(cancellationToken);
```

The problem I see with this approach is that `List<T>` implies an order of the elements, although the database does not guarantee an order as long as there is no `OrderBy()` in the query.

I know that this pattern is well established and it might be impossible to change it.
However to allow to better express that there is no order, I would like to see another extension method `ToHashSetAsync()`. 
Then one could write a query like this
```
HashSet<Blog> blogs = await context.Blogs.Where(b => something).ToHashSetAsync(cancellationToken);
```

This would better express that there is no ordering to all consumers of the `blogs` collection.

### PR DESCRIPTION
**Pull request by wertzui**
Added ToHashSetAsync<T> Extension methods on IQueryable<T>
- Added ToHashSetAsync<TSource>(this IQueryable<TSource> source, CancellationToken cancellationToken = default)
- Added ToHashSetAsync<TSource>(this IQueryable<TSource> source, IEqualityComparer<TSource>? comparer, CancellationToken cancellationToken = default)

Fixes #30033

### CODE REVIEW

**Comment by wertzui** [comment]
@dotnet-policy-service agree

**Review by ajcvickers — APPROVED** [review]

File: src/EFCore/Extensions/EntityFrameworkQueryableExtensions.cs [hunk 1]
```diff
@@ -2337,6 +2337,84 @@ await foreach (var element in source.AsAsyncEnumerable().WithCancellation(cancel
...
+    /// </returns>
+    /// <exception cref="ArgumentNullException"><paramref name="source" /> is <see langword="null" />.</exception>
+    /// <exception cref="OperationCanceledException">If the <see cref="CancellationToken" /> is canceled.</exception>
+    public static async Task<HashSet<TSource>> ToHashSetAsync<TSource>(
+        this IQueryable<TSource> source,
+        IEqualityComparer<TSource>? comparer,
+        CancellationToken cancellationToken = default)
+    {
+        var set = new HashSet<TSource>(comparer);
+        await foreach (var element in source.AsAsyncEnumerable().WithCancellation(cancellationToken).ConfigureAwait(false))
```
[the remark this answers is not in the document]
Comment (ajcvickers) [reply src/EFCore/Extensions/EntityFrameworkQueryableExtensions.cs:2408]: It is used in both `ToListAsync` and `ToDictionaryAsync`

**Comment by ajcvickers** [comment]
Thanks for the contribution!

**Comment by wertzui** [comment]
Thanks for approving it, I wanted to have `ToHashSetAsync` for quite some time.

**Pull request merged**